### PR TITLE
Add Validation loss for LoRA training

### PR DESCRIPTION
**Pull request by hinablue**
Base on https://github.com/kohya-ss/sd-scripts/pull/1165 @gesen2egee and https://github.com/kohya-ss/sd-scripts/pull/914 "Add validation loss" by @rockerBOO works.

And I only test on Lora training, DreamBooth and ControlNet training need more test.

截圖 2024-12-27 16 35 01

### CODE REVIEW

**Comment by rockerBOO** [comment]
instead of taking the code and making a commit of your own you should follow the full commit process so the credit and copyright is property accounted for. It's very rude to all those before it. The PR you took from also did it to my original code and it's really annoying as I get no credit

**Comment by hinablue** [comment]
> instead of taking the code and making a commit of your own you should follow the full commit process so the credit and copyright is property accounted for. It's very rude to all those before it. The PR you took from also did it to my original code and it's really annoying as I get no credit

Sorry for the late reply. I've included the original PR link with your information. If you think there is anything that needs to be corrected, please let me know. In fact, I have been following this PR for a long time, and I apologize for the trouble it has caused you.

**Comment by hinablue** [comment]
> I apologize for not being able to merge the previous validation loss PR. I would like to take advantage of the current holiday season to merge.
> 
> @hinablue What is the update from the previous PR?
> 
> @rockerBOO Is this PR already okay with you now?

I ref the https://github.com/kohya-ss/sd-scripts/pull/1165 @gesen2egee and rewrite it to work in the sd3 branch. Modified the calculation method of val loss after discussing with @gesen2egee (https://github.com/kohya-ss/sd-scripts/pull/1858/commits/62164e57925125ed6268983ffa441f1ffecc0e6d).

**Comment by hinablue** [comment]
Sorry to disturb everyone, I decided to close this PR. I will find another time to re-make the PR from @rockerBOO 's repo and send it to his sd3 branch.

**Comment by rockerBOO** [comment]
@hinablue i'm working to merge these together with the proper commits. Should be ready to review in a couple hours.

**Comment by rockerBOO** [comment]
@hinablue Also want to apologize as I was a little tough with my statement previously. 

I hope my combination of PR's will be adequate and I appreciate the work you did to fix any conflicts and update the code for the latest. Made the process pretty straight forward for me. I will make sure you get the appropriate credit on the official release, as well as your commits are preserved.
